datafusion/core/src/physical_plan/wrapper.rs

```diff
@@ -0,0 +1,160 @@
+    use arrow::datatypes::{DataType, Field};
+    use arrow_schema::{Schema, SchemaRef, SortOptions};
+    use arrow_array::{StringArray, DictionaryArray, Int32Array, types::Int32Type, RecordBatch, ArrayRef};
+
+    use super::*;
+
+    // Generate a record batch with a high cardinality dictionary field
+    fn generate_batch_with_cardinality(card: String) -> Result<RecordBatch, ArrowError> {
+        let schema = SchemaRef::new(Schema::new(vec![
+            Field::new("a_dict", DataType::Dictionary(Box::new(DataType::Int32), Box::new(DataType::Utf8)), false),
+            Field::new("b_prim", DataType::Int32, false),
+        ]));
+        
+        let col_a: ArrayRef;
+        if card == "high" {
+            // building column `a_dict`
+            let mut values_vector: Vec<String> = Vec::new();
+            for _i in 1..=20 {
+                values_vector.push(String::from(Uuid::new_v4().to_string()));
+            }
+            let values = StringArray::from(values_vector);
+
+            let mut keys_vector: Vec<i32> = Vec::new();
+            for _i in 1..=20 {
+                keys_vector.push(rand::thread_rng().gen_range(0..20));
+            }
+            let keys = Int32Array::from(keys_vector);
+            col_a = Arc::new(DictionaryArray::<Int32Type>::try_new(keys, Arc::new(values)).unwrap());
+        } else {
+            let values_vector = vec!["a", "b", "c"];
+            let values = StringArray::from(values_vector);
+            
+            let mut keys_vector: Vec<i32> = Vec::new();
+            for _i in 1..=20 {
+                keys_vector.push(rand::thread_rng().gen_range(0..2));
+            }
+            let keys = Int32Array::from(keys_vector);
+            col_a = Arc::new(DictionaryArray::<Int32Type>::try_new(keys, Arc::new(values)).unwrap());
+        }
+
+        // building column `b_prim`
+        let mut values: Vec<i32> = Vec::new();
+            for _i in 1..=20 {
+                values.push(rand::thread_rng().gen_range(0..20));
+            }
+        let col_b: ArrayRef = Arc::new(Int32Array::from(values));
+
+        // building the record batch
+        RecordBatch::try_from_iter(vec![("a_dict", col_a), ("b_prim", col_b)])
+    }
+
+    #[tokio::test]
+    async fn test_with_high_card() {
+        let batch = generate_batch_with_cardinality(String::from("high")).unwrap();
+        let sort_fields = vec![
+            arrow::row::SortField::new_with_options(DataType::Dictionary(Box::new(DataType::Int32), Box::new(DataType::Utf8)), SortOptions::default()),
+            arrow::row::SortField::new_with_options(DataType::Int32, SortOptions::default())
+        ];
+        let mut converter = CardinalityAwareRowConverter::new(sort_fields.clone()).unwrap();
+        let rows = converter.convert_columns(&batch.columns()).unwrap();        
+        let converted_batch = converter.convert_rows(&rows).unwrap();
+        assert_eq!(converted_batch[0].data_type(), &DataType::Utf8);
```

Comment:
it might be worth a comment here explaining that we expect the type to be changed